### PR TITLE
fix: make mda engine exit gracefully after an exception

### PR DESCRIPTION
**Pull request by ianhi**
Fixes https://github.com/tlambert03/pymmcore-plus/issues/130

### CODE REVIEW

File: pymmcore_plus/mda/_engine.py [hunk 1]
```diff
@@ -248,20 +248,31 @@ def run(self, sequence: MDASequence) -> None:
...
-        self._finish_run(sequence)
+                self._events.frameReady.emit(img, event)
+            self._finish_run(sequence)
+        except Exception as e:  # noqa E722
+            # clean up so future MDAs can be run
+            self._canceled = False
+            self._running = False
+            try:
+                self._finish_run(sequence)
+            except:  # noqa E722
```
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:275]:
```suggestion
            except Exception:
```

File: pymmcore_plus/mda/_engine.py [hunk 2]
```diff
@@ -248,20 +248,31 @@ def run(self, sequence: MDASequence) -> None:
...
+                # ignore this one and raise the first failure
+                pass
```
Comment (tlambert03) [inline pymmcore_plus/mda/_engine.py:277]:
```suggestion
            with contextlib.suppress(Exception):
                self._finish_run(sequence)
```

Comment (ianhi) [reply pymmcore_plus/mda/_engine.py]: pretty slick.

File: pymmcore_plus/mda/_engine.py [hunk 3]
```diff
@@ -248,20 +249,28 @@ def run(self, sequence: MDASequence) -> None:
...
-            self._events.frameReady.emit(img, event)
+                self._events.frameReady.emit(img, event)
+        except Exception as e:  # noqa E722
+            # clean up so future MDAs can be run
+            self._canceled = False
+            self._running = False
+            with contextlib.suppress(Exception):
+                self._finish_run(sequence)
+            raise e
         self._finish_run(sequence)
```
Comment (ianhi) [inline pymmcore_plus/mda/_engine.py:276]: moved this to below the `try/except` so that if it throws an error we don't then try to throw an error twice.

**Review by tlambert03 — APPROVED** [review]
Lgtm!

**Pull request merged**